storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -239,6 +246,10 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
         markFailedTopology(topologySubmitter, cluster, td, "Failed to schedule within " + maxSchedulingAttempts + " attempts");
     }
 
+    public Set<String> getEvictedTopologies() {
```

Comment:
I am seeing a possible race condition here. Since the `evictedTopologies` will be kept updated during scheduling, any function call to `getEvictedTopologies` doesn't guarantee to return a useful result.  This `evictedTopologies` includes every topology that is evicted during the scheduling process until the point when `getEvictedTopologies` is called.  It doesn't seem useful. 

To know what topologies are being evicted, we already have
```
LOG.debug("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict,
                                    topologyEvict.getTopologySubmitter());
```
code in place. We can change it to `warn` if we want it to show up in the regular logs.  

@kishorvpatil  what do you think?

Reply:
In that case this method should return a copy "return new HashSet<>(this.evictedTopologies);"

Reply:
Yes that's right. We should return a unmodifiableCollection to solve the race condition. I think my question is mostly on how do we want to use this `evictedTopologies` information in regular code except unit test?

Reply:
As a member of SchedulingResult this information can be more useful.
But looks like the intent is to keep track of topologies evicted as part of each topo scheduled, in which case we need a map i.e. topoId -> toposEvicted